lazy load types

crissi opened the pull request:
This only works if types are added as "key" => Type::class in config, since bootTypes::bootTypes will instanciate a new entry of all types if it can not associate the name directly. See addType in GraphQL.php

### Links
- Related to https://github.com/rebing/graphql-laravel/issues/390

mfn commented on src/GraphQL.php:
```diff
@@ -55,33 +55,13 @@ public function schema($schema = null): Schema
-        foreach ($this->getTypes() as $name => $type) {
-            $this->type($name);
-        }
-
```
I forgot to mention that yesterday I also removed this code because I also had the feeling it wasn't doing much 😄

crissi replied:
Tested on a live system and failed.

The type loader receives the name from the Types name-attribute wich can be different from the alias used in config. That mean we have to check every file for the name.

One way, could be to make a artisan graphql cache command that mapped the type names to the classes so it was easily found.

But I might be missing something...

mfn replied:
> One way, could be to make a artisan graphql cache command that mapped the type names to the classes so it was easily found.

This direction seem appropriate to explore!

Basically we need to figure out what and how can we cache it and then make a config entry for this.

With such a thing in place, if you deploy code (changes to GraphQL type system in some way) and don't clear the cache it would probably break. Which is _fine_ because the recommendation _is_ to clear the cache after deployment.

mfn commented on src/GraphQL.php:
```diff
@@ -94,11 +73,37 @@ public function schema($schema = null): Schema
+            'mutation' => !empty($schemaMutation) ? $mutation : null,
+            'subscription' => !empty($schemaSubscription) ? $subscription : null,
+            'typeLoader' => function ($name) {
+                $keyName = array_search($name, $this->typesInstances);
```
We should try to avoid `array_search` as it has linear performance characteristics.

crissi replied:
> We should try to avoid `array_search` as it has linear performance characteristics.

waiting to change this, to after I have done some benchmarking

mfn replied:
Do you have any results?

mfn commented:
I've rebased with master (fixed styleci, adapted phpstan) where I just added test for file uploads.

Really just a coincidence, but these tests now break :|

(guess phpstan was right after all about the return value from `array_search` and I should not have put it into `ignoreErrors`?)

crissi commented:
> I've rebased with master (fixed styleci, adapted phpstan) where I just added test for file uploads.
> 
> Really just a coincidence, but these tests now break :|
> 
> (guess phpstan was right after all about the return value from `array_search` and I should not have put it into `ignoreErrors`?)

so far I don't know how to fix the upload type (custom scalar types), the name I get in is Upload, but I have know Idea where to look for the class name

crissi commented:
> Do we have any recent performance numbers (is it working 😄 )?

nope, might get to benchmark a bit tomorrow

mfn commented:
Rebased with master, applied StyleCI fixes and adapted phpstan.

Btw: StyleCI violations are easy to fix => click on the link => download the raw diff => apply patch (I use PhpStorm which has a feaure "apply patch from clipboard")

mfn commented on src/Support/UploadType.php:
```diff
@@ -61,4 +63,9 @@ public static function getInstance()
+
+    public function toType(): Type
+    {
+        return self::getInstance();
```
I suggest ~self~ `static`

mfn replied:
Obsolete, handled in master now (and differently)

mfn commented on src/GraphQLServiceProvider.php:
```diff
@@ -77,6 +78,8 @@ protected function bootPublishes(): void
     {
         $configTypes = config('graphql.types');
+
+        $configTypes['Upload'] = UploadType::class;
```
Ah, I see!

Now instead of using this:
```php
    public function args(): array
    {
        return [
            'files' => [
                'type' => Type::listOf(UploadType::getInstance()),
            ],
        ];
    }
```
we can use
```php
                'type' => Type::listOf(GraphQL::type('Upload')),
```
which also is more idiomatic how things are defined.

However my feeling is we should not add this hardcoded: not everyone might want to use the upload capability.

In fact, it's the same with the upload middleware (always enabled).

I feel there needs to be some refactoring for the middleware _and_ the `Upload` type should be added by the user manually if they want to use it (effectively making this a documentation issue).

FTR I'm also toying with the idea removing the whole upload middleware from this package in its current form and rather use https://github.com/Ecodev/graphql-upload and have an optionally enabled adapter/bridge.

mfn replied:
> I feel there needs to be some refactoring for the middleware and the Upload type should be added by the user manually if they want to use it (effectively making this a documentation issue).

FTR I'm talking about _outside_ this PR!

mfn replied:
For the UploadType change, see https://github.com/rebing/graphql-laravel/pull/401

mfn replied:
The UploadType change was merged, thus I've rebased this PR _and_ removed the hardcoded UploadType, as it isn't necessary anymore.

crissi commented:
thought I fixed it with the names being diffent in class queries, must investigate again

mfn commented:
Can you elaborate a bit on the last commit/change?

I'm just wondering why no test broke before that change (was under the impression these sort of things were covered).

crissi commented:
No, there was no types in the tests covering types that had a different config name than it equalient class type name

Få Outlook til iOS<https://aka.ms/o0ukef>
________________________________
Fra: Markus Podar <notifications@github.com>
Sendt: Thursday, July 11, 2019 7:59:33 PM
Til: rebing/graphql-laravel
Cc: Christian; Author
Emne: Re: [rebing/graphql-laravel] [WIP] lazy load types (#394)


Can you elaborate a bit on the last commit/change?

I'm just wondering why no test broke before that change (was under the impression these sort of things were covered).

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/rebing/graphql-laravel/pull/394?email_source=notifications&email_token=AAXPDYSDNH6NCSOZBBL3QGDP65YILA5CNFSM4H6ZFWD2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZXP4ZQ#issuecomment-510590566>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AAXPDYXJ32MSZBYGHAAIYADP65YILANCNFSM4H6ZFWDQ>.

crissi commented:
So right now it is alot faster if the config alias and the class name alias matches, but we have no efficient way to check that in runtime. So we can not rely on that!

So the only step  I can come up with is to make a cache artisan command. I might get to work on that in the weekend.

Do you have any other suggestions or comments?

mfn commented:
> No, there was no types in the tests covering types that had a different config name than it equivalentclass type name

😱

In that case, thanks!

> So right now it is alot faster if the config alias and the class name alias matches, but we have no efficient way to check that in runtime.

What is even the use case for this I wonder?

TBH never understood this aspect and it certainly would help to get rid of it … _thinking out loud_

> Do you have any other suggestions or comments?

My understanding is that after your hard work this is getting into its final stages, so I'll start a more in-depth review (I hope soon).

crissi commented:
I will probably clean it up a bit and do the cache part in A new PR

Få Outlook til iOS<https://aka.ms/o0ukef>
________________________________
Fra: Markus Podar <notifications@github.com>
Sendt: Friday, July 12, 2019 6:15:00 AM
Til: rebing/graphql-laravel
Cc: Christian; Author
Emne: Re: [rebing/graphql-laravel] [WIP] lazy load types (#394)


No, there was no types in the tests covering types that had a different config name than it equivalentclass type name

😱

In that case, thanks!

So right now it is alot faster if the config alias and the class name alias matches, but we have no efficient way to check that in runtime.

What is even the use case for this I wonder?

TBH never understood this aspect and it certainly would help to get rid of it … thinking out loud

Do you have any other suggestions or comments?

My understanding is that after your hard work this is getting into its final stages, so I'll start a more in-depth review (I hope soon).

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/rebing/graphql-laravel/pull/394?email_source=notifications&email_token=AAXPDYTFNEOYXTDW6SXQI4LP7AAMJA5CNFSM4H6ZFWD2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZYT3RA#issuecomment-510737860>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AAXPDYUDAJKDW73K5VEN4ULP7AAMJANCNFSM4H6ZFWDQ>.

mfn commented:
Sounds good to me!

crissi commented:
> > So right now it is alot faster if the config alias and the class name alias matches, but we have no efficient way to check that in runtime.
> 
> What is even the use case for this I wonder?
> 

Had the same thought. In my application I I have made it work with class names instead. So I'dont have to also remember type aliases.

![image](https://user-images.githubusercontent.com/3076578/61123246-28f13900-a4a4-11e9-8fb9-4ec5edbcab1e.png)

mfn commented:
Interesting!

I made a similar approach but decided to for `TYPE` constants and use them when referencing the types:
![image](https://user-images.githubusercontent.com/87493/61128033-395be080-a4b1-11e9-88fa-c6579a7f66b0.png)

crissi commented:
I am thinking maybe this typeloader should be a switch people can toggle on, if they are using types "correctly" by having the same alias in config and in names? Will make us not fight the underlying framework and make it just work (FAST). Will of cause need some docs to explain how it should be set up

mfn commented:
I don't want to jump the gun but I rather have one clear system than two to maintain.

Sorry I haven't yet found enough time for it, I haven't forgotten it!

mfn commented:
Closing in favor of https://github.com/rebing/graphql-laravel/pull/405
